### PR TITLE
fix(redirect): handle redirects explicitly 

### LINKED ISSUE #26: can't fetch url which will redirect to another url with different protocol
```js
const fetch = require('make-fetch-happen');
fetch('http://registry.npm.taobao.org/istanbul-api/download/istanbul-api-1.1.8.tgz').catch(err => console.log(err.stack));
```

=> 

```
Error: Protocol "https:" not supported. Expected "http:"
    at new ClientRequest (_http_client.js:54:11)
    at Object.exports.request (http.js:31:10)
    at exports.request (https.js:199:15)
    at fetch.Promise (/Users/deadhorse/git/github.com/eggjs/egg/node_modules/._node-fetch-npm@2.0.0@node-fetch-npm/src/index.js:54:17)
    at fetch (/Users/deadhorse/git/github.com/eggjs/egg/node_modules/._node-fetch-npm@2.0.0@node-fetch-npm/src/index.js:41:10)
    at ClientRequest.req.on.res (/Users/deadhorse/git/github.com/eggjs/egg/node_modules/._node-fetch-npm@2.0.0@node-fetch-npm/src/index.js:114:17)
    at emitOne (events.js:96:13)
    at ClientRequest.emit (events.js:188:7)
    at HTTPParser.parserOnIncomingClient (_http_client.js:473:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23)
```

```bash
curl -I http://registry.npm.taobao.org/istanbul-api/download/istanbul-api-1.1.8.tgz

HTTP/1.1 302 Found
Server: nginx/1.4.6 (Ubuntu)
Date: Wed, 10 May 2017 05:11:05 GMT
Connection: keep-alive
X-Current-Requests: 1
Location: https://cdn.npm.taobao.org/istanbul-api/-/istanbul-api-1.1.8.tgz
X-ReadTime: 4
```

node-fetch will use the same agent when request the tow different protocol url.

### PR DESCRIPTION
**Pull request by nlkluth**
Fixes #26

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -357,7 +360,53 @@ function remoteFetch (uri, opts) {
...
+          if (opts.redirect === 'error') {
+            throw new Error(`redirect mode is set to error: ${uri}`, 'no-redirect')
+          }
+
+          if (!res.headers.get('location')) {
+            throw new Error(`redirect location header missing at: ${uri}`, 'invalid-redirect')
+          }
+
+          if (req.counter >= req.follow) {
+            throw new Error(`maximum redirect reached at: ${uri}`, 'max-redirect')
```
Comment (zkat) [inline index.js]: What does this second argument do? The convention so far has been to use a `code` property on error objects to have single codes that people can use.

**Review by zkat — CHANGES_REQUESTED** [review]
This is really great! Thank you!

Only one small change request involving the errors themselves, but the lest looks great, is super thorough, and I love the tests. They covered pretty much anything I was keeping an eye out for. 👏🏼

File: index.js (hunk 1 above)
Comment (nlkluth) [reply index.js]: ah, this is my mistake. I'll get the change going thanks! Are there a set of codes to go off of?

Comment (zkat) [reply index.js]: nah. My main suggestion would be to use the convention I've been using for other libraries, so `E<SOMECODE>` in uppercase. Unix-style.

Comment (nlkluth) [reply index.js]: Gave it a go - but happy to make changes :)

**Review by zkat — APPROVED** [review]

**Comment by zkat** [comment]
💃

**Comment by zkat** [comment]
This has been released as part of `2.4.4`. Thanks so much for this. It's awesome work, and *incredibly* helpful, and I really appreciate your time <3

**Comment by nlkluth** [comment]
Thanks! 🙌

**Pull request merged**